core-primitives/stf-executor/src/executor.rs

```diff
@@ -232,9 +232,11 @@ where
 		// Update parentchain block on all states.
 		let shards = self.state_handler.list_shards()?;
 		for shard_id in shards {
-			let (_, mut state) = self.state_handler.load_for_mutation(&shard_id)?;
+			let (state_lock, mut state) = self.state_handler.load_for_mutation(&shard_id)?;
 			if let Err(e) = Stf::update_parentchain_block(&mut state, header.clone()) {
 				error!("Could not update parentchain block. {:?}: {:?}", shard_id, e)
+			} else {
+				self.state_handler.write(state, state_lock, &shard_id)?;
```

Comment:
Oof, yeah that seems like a blunder. Did I introduce this with my refactoring?

Reply:
Don't think so. I think it was simply overlooked by all of us.

Reply:
We should probably use `match` then:
```
match Stf::update_parentchain_block(...) {
   Err(e) => {},
   Ok(_) => {
      self.state_handler.write()
   }
}
```

Reply:
I'm not against that, but what's your reasoning? For me both of them are equal.

Reply:
It's more about style, I think the Rust guidelines say, if you only need one pattern match, then use `let Err(e) = `. Otherwise use `match` to deal with all the cases. It's better readable than all the `if else` and less error prone (although in the case of `Result` that shouldn't be too much of a problem)

Reply:
Although I have to admit the difference is minor ☺️

Reply:
Uff, we need more tests....